performance/class-mime-types-caching.php

```diff
@@ -0,0 +1,163 @@
+<?php
+/**
+ * MIME Types Caching.
+ *
+ * @package vip-performance
+ */
+
+// phpcs:disable WordPress.DB.DirectDatabaseQuery.DirectQuery
+
+namespace Automattic\VIP\Performance;
+
+/**
+ * MIME Types Caching class.
+ */
+class Mime_Types_Caching {
+	public const MINIMUM_WORDPRESS_VERSION  = '6.4';
+	public const CACHE_GROUP                = 'mime_types';
+	public const MIME_TYPES_CACHE_KEY       = 'vip_mime_types';
+	public const MAX_POSTS_TO_QUERY_DEFAULT = 500000;
+
+	/**
+	 * Class initialization.
+	 */
+	public static function init() {
+		global $wp_version;
+
+		if ( isset( $wp_version ) &&
+			version_compare( $wp_version, self::MINIMUM_WORDPRESS_VERSION, '>=' ) ) {
+			static::enable_post_mime_types_caching();
+		}
+	}
+
+	/**
+	 * Enable MIME type caching.
+	 */
+	private static function enable_post_mime_types_caching() {
+		add_filter( 'pre_get_available_post_mime_types', array( __CLASS__, 'get_cached_post_mime_types' ), 10, 2 );
+		add_action( 'add_attachment', array( __CLASS__, 'add_mime_type_to_cache' ), 10, 1 );
+		add_action( 'attachment_updated', array( __CLASS__, 'update_mime_types_cache' ), 10, 3 );
+		add_action( 'delete_attachment', array( __CLASS__, 'delete_mime_types_cache' ), 10 );
+	}
+
+	/**
+	 * Get cached results for get_available_post_mime_types() to avoid a query on every page load.
+	 *
+	 * @param string[]|null $filtered_mime_types An array of MIME types. Default null.
+	 * @param string        $type                The post type name.
+	 * @return array An array of MIME types.
+	 */
+	public static function get_cached_post_mime_types( $filtered_mime_types, $type ) {
+		if ( 'attachment' !== $type ) {
+			return $filtered_mime_types;
+		}
+
+		$mime_types = wp_cache_get( self::MIME_TYPES_CACHE_KEY, self::CACHE_GROUP );
+
+		if ( false === $mime_types ) {
+
+			/**
+			 * Filters the max number of posts to query for dynamic MIME type caching.
+			 *
+			 * @param bool $max_posts_to_query Max number of posts to query.
+			 */
+			$max_posts_to_query = apply_filters( 'vip_max_posts_to_query_for_mime_type_caching', self::MAX_POSTS_TO_QUERY_DEFAULT );
```

Comment:
```suggestion
			$max_posts_to_query = absint( apply_filters( 'vip_max_posts_to_query_for_mime_type_caching', self::MAX_POSTS_TO_QUERY_DEFAULT ) );
```

Then we can remove the check on the off-chance that it evaluates to less than 0.

Reply:
I structured it like that to have a way to skip the query in case a user always wanted to use the defaults. Then they can just set `MAX_POSTS_TO_QUERY_DEFAULT` to zero a negative value, like `-1`, and we can skip the `COUNT()` query altogether.

This approach could be helpful for users who wouldn't ever be near the threshold for running the full query in the DB, like someone with millions of attachments.

It should probably be documented, but does it make sense?

Reply:
Makes sense. Can you add a note in that filter PHPDoc?

Reply:
Added with https://github.com/Automattic/vip-go-mu-plugins/pull/4953/commits/c654f6bc77fed90f4cec6fa1bba6bc8369846889.